File: packages/rtc_audio_effects/android/src/main/kotlin/ly/plugins/rtc_audio_effects/BaseEffectCoreEventHandler.kt
package ly.plugins.rtc_audio_effects

import android.os.Handler
import android.os.Looper
import com.voice.core.RtcEffectCore
import com.voice.core.RtcEffectEventHandler
import ly.plugins.rtc_audio_effects.BaseEffectConstants.TAG
import io.flutter.plugin.common.EventChannel.EventSink

object BaseEffectCoreEventHandler {

  var sink: EventSink? = null

  private val mUIHandler: Handler by lazy {
    Handler(Looper.getMainLooper())
  }

  var engine: RtcEffectCore? = null

  private fun guardSink(): Boolean {
    if (sink == null) {
      BaseEffectUtils.logE(TAG, "FlutterEventSink is null")
      return true
    }
    return false
  }

  val eventHandler = object : RtcEffectEventHandler {
    override fun onAudioRouteChanged(routing: Int) {
      BaseEffectUtils.logD(TAG, "eventHandler onAudioRouteChanged routing->$routing")
      if (guardSink()) {
        return
      }
      val map = mutableMapOf<String, Any>()
      map["method"] = "onAudioRouteChanged"
      map["routing"] = routing
      mUIHandler.post {
        sink!!.success(map)
      }
    }

    override fun onStateChanged(state: Int) {
      BaseEffectUtils.logD(TAG, "eventHandler onStateChanged->$state")
      if (guardSink()) {
        return
      }

      val map = mutableMapOf<String, Any>()
      map["method"] = "onStateChanged"
      map["state"] = state
      mUIHandler.post {
        sink!!.success(map)
      }
    }

    override fun onKMarkSliceScoreUpdate(beginPosMs: Int, endPosMs: Int, pitch: Float, score: Float) {

      if (guardSink()) {
        return
      }
      val map = mutableMapOf<String, Any>()
      map["method"] = "onKMarkSliceScoreUpdate"
      map["beginPosMs"] = beginPosMs
      map["endPosMs"] = endPosMs
      map["pitch"] = pitch
      map["score"] = score
      mUIHandler.post {
        sink!!.success(map)
      }
    }

    override fun onKMarkSentenceScoreUpdate(index: Int, currentScore: Float, totalScore: Float) {

      if (guardSink()) {
        return
      }
      val map = mutableMapOf<String, Any>()
      map["method"] = "onKMarkSentenceScoreUpdate"
      map["index"] = index
      map["currentScore"] = currentScore
      map["totalScore"] = totalScore
      mUIHandler.post {
        sink!!.success(map)
      }
    }

    override fun onRecordAudioFrame(sampleRate: Int, channels: Int, samples: Int, data: ByteArray?) {
      if (guardSink()) {
        return
      }
      val map = mutableMapOf<String, Any?>()
      map["method"] = "onRecordAudioFrame"
      map["sampleRate"] = sampleRate
      map["channels"] = channels
      map["samples"] = samples
      map["data"] = data
      mUIHandler.post {
        sink!!.success(map)
      }
    }

    override fun onProductProgress(progress: Float) {
      if (guardSink()) {
        return
      }
      val map = mutableMapOf<String, Any>()
      map["method"] = "onProductProgress"
      map["progress"] = progress
      mUIHandler.post {
        sink!!.success(map)
      }
    }

    override fun onAudioVolumeIndication(volume: Int, timeMs: Long){
      if (guardSink()) {
        return
      }
      val map = mutableMapOf<String, Any>()
      map["method"] = "onAudioVolumeIndication"
      map["volume"] = volume
      map["timeMs"] = timeMs
      mUIHandler.post {
        sink!!.success(map)
      }
    }
  }
}

File: packages/rtc_audio_effects/ios/Classes/BaseEffectCoreEventHandler.m
//
//  BaseEffectCoreEventHandler.m
//  base_effect_core
//
//  Created by eagle on 2022/2/22.
//

#import "BaseEffectCoreEventHandler.h"

@implementation BaseEffectCoreEventHandler

- (BOOL)guardSink{
    if (self.eventSink == nil) {
        return true;
    }
    return false;
}

- (void)audioRouteChanged:(int)routing {
    if ([self guardSink]) {
        return;
    }
    
    NSMutableDictionary *dic = [NSMutableDictionary new];
    [dic setObject:@"onAudioRouteChanged" forKey:@"method"];
    [dic setObject:[NSNumber numberWithInt:routing] forKey:@"routing"];

    dispatch_async(dispatch_get_main_queue(), ^{
        self.eventSink(dic);
    });
}

- (void)stateChanged:(int)state {
    if ([self guardSink]) {
        return;
    }
    
    NSMutableDictionary *dic = [NSMutableDictionary new];
    [dic setObject:@"onStateChanged" forKey:@"method"];
    [dic setObject:[NSNumber numberWithInt:state] forKey:@"state"];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        self.eventSink(dic);
    });
}

- (void)sliceScoreUpdate:(int)beginPosMs endPosMs:(int)endPosMs pitch:(float)pitch score:(float)score {
    if ([self guardSink]) {
        return;
    }
    
    NSMutableDictionary *dic = [NSMutableDictionary new];
    [dic setObject:@"onKMarkSliceScoreUpdate" forKey:@"method"];
    [dic setObject:[NSNumber numberWithInt:beginPosMs] forKey:@"beginPosMs"];
    [dic setObject:[NSNumber numberWithInt:endPosMs] forKey:@"endPosMs"];
    [dic setObject:[NSNumber numberWithFloat:pitch] forKey:@"pitch"];
    [dic setObject:[NSNumber numberWithFloat:score] forKey:@"score"];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        self.eventSink(dic);
    });
}

- (void)sentenceScoreUpdate:(int)index currentScore:(float)currentScore totalScore:(float)totalScore {
    if ([self guardSink]) {
        return;
    }
        
    NSMutableDictionary *dic = [NSMutableDictionary new];
    [dic setObject:@"onKMarkSentenceScoreUpdate" forKey:@"method"];
    [dic setObject:[NSNumber numberWithInt:index] forKey:@"index"];
    [dic setObject:[NSNumber numberWithFloat:currentScore] forKey:@"currentScore"];
    [dic setObject:[NSNumber numberWithFloat:totalScore] forKey:@"totalScore"];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        self.eventSink(dic);
    });
}

- (void)recordAudioFrame:(int)sampleRate channels:(int)channels samples:(int)samples data:(int16_t * _Nullable)data {
    NSMutableDictionary *dic = [NSMutableDictionary new];
    [dic setObject:@"onRecordAudioFrame" forKey:@"method"];
    [dic setObject:[NSNumber numberWithInt:sampleRate] forKey:@"sampleRate"];
    [dic setObject:[NSNumber numberWithInt:channels] forKey:@"channels"];
    [dic setObject:[NSNumber numberWithInt:samples] forKey:@"samples"];

    NSData *ocData = [NSData dataWithBytes:data length:sizeof(int16_t)];
    FlutterStandardTypedData *resData = [FlutterStandardTypedData typedDataWithBytes:ocData];
    [dic setObject:resData forKey:@"data"];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        self.eventSink(dic);
    });
}

- (void)productProgress:(float)progress {
    NSMutableDictionary *dic = [NSMutableDictionary new];
    [dic setObject:@"onProductProgress" forKey:@"method"];
    [dic setObject:[NSNumber numberWithFloat:progress] forKey:@"progress"];

    dispatch_async(dispatch_get_main_queue(), ^{
        self.eventSink(dic);
    });
}

- (void)audioVolumeIndication:(int)volume time:(int64_t)time{
    NSMutableDictionary *dic = [NSMutableDictionary new];
    [dic setObject:@"onAudioVolumeIndication" forKey:@"method"];
    [dic setObject:[NSNumber numberWithInt:volume] forKey:@"volume"];
    [dic setObject:[NSNumber numberWithInt:time] forKey:@"timeMs"];

    dispatch_async(dispatch_get_main_queue(), ^{
        self.eventSink(dic);
    });
}

@end
